src/handlers/psp.ts

```diff
@@ -169,29 +168,70 @@ const initializeState = (config: Config): State => {
 const executeApiCalls = async (state: State): Promise<State> => {
   const { config, baseLogOptions, groupedSubscriptions } = state;
 
-  const apiValuePromises = groupedSubscriptions.map(({ subscriptions, template, endpoint }) => {
+  const wait = (ms: number) => new Promise((res) => setTimeout(res, ms));
+  const doRequestWithRetries = async (
+    fn: () => Promise<CallApiResult>,
+    minDelay = 1_000,
+    maxDelay = 5_000
+  ): Promise<CallApiResult> => {
+    try {
+      return await fn();
+    } catch (e) {
+      await wait(Math.floor(Math.random() * (maxDelay - minDelay + 1) + minDelay));
+      return await doRequestWithRetries(fn, minDelay, maxDelay);
+    }
+  };
+
+  let hasApiCallWrapperTimedOut = false;
+  const responses: CallApiResult[] = [];
+  const errorLogs: utils.PendingLog[] = [];
+  const apiValuePromises = groupedSubscriptions.map(async ({ subscriptions, template, endpoint }) => {
     const apiCallParameters = abi.decode(template.templateParameters);
-    return go(
-      async () => {
-        const [logs, data] = await callApi(config, endpoint, apiCallParameters);
-        return [logs, { templateId: template.id, apiValue: data, subscriptions }] as node.LogsData<{
-          templateId: string;
-          apiValue: ethers.BigNumber | null;
-          subscriptions: Id<Subscription>[];
-        }>;
-      },
-      { timeoutMs: TIMEOUT_MS, retries: RETRIES }
-    );
+
+    let hasApiCallTimedOut = false;
+    const result = await doRequestWithRetries(async () => {
+      if (hasApiCallWrapperTimedOut) {
+        return [[], { templateId: template.id, apiValue: null, subscriptions }] as CallApiResult;
+      }
+      const result = await promise.go(
+        async () => {
+          try {
+            return await callApi(config, endpoint, apiCallParameters);
+          } catch (err) {
+            if (hasApiCallTimedOut) {
+              return [
+                [utils.logger.pend('ERROR', err instanceof Error ? err.message : String(err))],
+                { templateId: template.id, apiValue: null, subscriptions },
+              ] as CallApiResult;
+            }
```

Comment:
I still think this doesn't work in all cases. Imagine that the first attempt will timeout, meaning that the `callApi` is still running. We set the `hasApiCallTimedOut` to true and start the second attempt. If that fails (within the timeout) it will be falsely considered as a failure of the first attempt after the timeout. Feel free to correct me if I'm wrong.

Reply:
The scenario makes sense. Let me try to reproduce it locally and see what solution can be applied if true.

Reply:
I don't think it's actually fixable in the current state, that's why I want to address it once we have the library fixed.

Reply:
I see. The thing is that I'm almost 💯 sure that I did test that scenario and it was working but the logic of what you said makes sense so I want to double check if it's an issue or not.

Reply:
Ok you were right @amarthadan. I just tested it and after timeout, if the next retry fails then we return null and stop retrying even though we want to keep retrying until we hit the 40 sec limit for all API requests.

I might be missing something but I think the problem is that I'm initializing the `hasApiCallTimedOut` outside the retry callback function. If we move it inside in between lines 195-196 then we would throw after the first timeout and on the next retry the flag would be false and keep retrying. I did a test of that change and it seems to be working but I might be missing other edge cases...

Reply:
Done! Ran a few tests with random timeouts + random failures and it seems to behave as expected